### ISSUE TITLE
Second call to templated method fails to resolve type parameter

### ISSUE THREAD

**Issue by weirdan** [vimeo/psalm#741]
https://getpsalm.org/r/a88b3506b8
```php
<?php
  namespace NS;

  use Closure;

  /**
   * @template TKey
   * @template TValue
   */
  class ArrayCollection {

    /** @var array<TKey,TValue> */
    private $data;

    /** @param array<TKey,TValue> $data */
    public function __construct(array $data) {
      $this->data = $data;
    }

    /**
     * @template T
     * @param Closure(TValue):T $func
     * @return ArrayCollection<TKey,T>
     */    
    public function map(Closure $func) {
      return new static(array_map($func, $this->data));
    }
  }

  class Item {}
  /**
   * @param ArrayCollection<mixed,Item> $i
   */
  function takesCollectionOfItems(ArrayCollection $i): void {}

  $c = new ArrayCollection([ new Item ]);
  takesCollectionOfItems($c);
  takesCollectionOfItems($c->map(function(Item $i): Item { return $i;}));
  takesCollectionOfItems($c->map(function(Item $i): Item { return $i;}));
```
Here, second call to `takesCollectionOfItems()` fails because psalm thinks the result of `map` is literal `ArrayCollection<int,T>`, where `T` should have been actually resolved to `Item`. First call works fine.

**Comment by weirdan** [comment]
/ping @muglug 
I see this is not categorized, unlike most other issues. Have you had a chance to take a look at this?

**Comment by muglug** [comment]
Rookie error - missing `__clone` in `TCallable`/`Fn` (Closure) types

**Comment by weirdan** [comment]
May have also fixed #732 (not the recursive types, of course, but the return type for that recursive generic function is now inferred correctly).

**Comment by muglug** [comment]
Yup, seems to be good
